[linux_platform] Implement a bunch of missing functions necessaries to make Xenia build correctly under Linux

RodoMa92 opened the pull request:
With this part fixed and a function re adapted from the original codebase to fix the different signature, we should have a compiling Linux build now.

A lot of threading_posix.cc has been cleaned up as recommended from clang-lint since I had some issues stemming from missing implementation on destructor and properly overridden destructor classes causing pthread_kill causing double exceptions and killing the process as a result.

Gliniak commented on src/xenia/base/filesystem_posix.cc:
```diff
@@ -240,7 +247,30 @@ std::vector<FileInfo> ListFiles(const std::filesystem::path& path) {
+  return std::move(result);
+}
+
+std::vector<FileInfo> ListDirectories(const std::filesystem::path& path) {
```
This part (``ListDirectories`` and ``FilterByName``) looks like might be directly in filesystem.cc. That way there will be only one implementation

RodoMa92 replied:
Done

Gliniak commented:
Remove implementation for ListDirectories and FilterByName from filesystem_win.cc

RodoMa92 commented:
> Remove implementation for ListDirectories and FilterByName from filesystem_win.cc

Doh, dumb me. Fixing it now.

RodoMa92 commented:
Should be clean now.

Gliniak commented:
Linter remains >_<

RodoMa92 commented:
It "should" be clean now. This also should be the first build to compile correctly with Linux, once the FFMPEG submodule linker fixes are in.

RodoMa92 committed:
[linux_platform] Implement a bunch of missing functions necessaries to make Xenia build correctly under Linux

With this part fixed and a function readapted from the original codebase to fix the different signature, we should have a compiling Linux build now.

RodoMa92 committed:
[filesystem] Moved generic methods to platform independent code

RodoMa92 committed:
[build] Add missing linker libraries to xenia-ui

Shoegzer commented:
Linux builds are still failing here, is this normal and expected still?

Test system info:
* OS: Linux Mint 22.1 (Ubuntu 22.04 base) / kernel 6.8
* CPU: Intel i7-10870H
* Build tested: 00202f938dde50dcd0a4040159a777e6f6fd4a2f
* Build log: [build.log](https://github.com/user-attachments/files/18481541/build.log)

RodoMa92 commented:
> Linux builds are still failing here, is this normal and expected still?
> 
> Test system info:
> * OS: Linux Mint 22.1 (Ubuntu 22.04 base) / kernel 6.8
> * CPU: Intel i7-10870H
> * Build tested: 00202f938dde50dcd0a4040159a777e6f6fd4a2f
> * Build log: [build.log](https://github.com/user-attachments/files/18481541/build.log)

You are missing premake as a system provided binary like me and I haven't tested building from source, which apparently fails.

Install premake, cmake, ninja and llvm/clang and it should build fine.

Shoegzer commented:
Thanks.  Since the Ubuntu repo only includes premake4, the build should be using premake5 from third party.  From `.gitmodules` it seems the project pulls from the latest upstream premake5, however the build script shows it's not cloning to build:
```
error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
error: 6059 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/premake/premake-core.git' into submodule path '/home/testing/Desktop/xenia/xenia-canary/third_party/premake-core' failed
```
Otherwise, `cmake 3.28.3`, `ninja 1.11.1` and clang/llvm `18.1.3` are installed.

RodoMa92 commented:
> Thanks. Since the Ubuntu repo only includes premake4, the build should be using premake5 from third party. From `.gitmodules` it seems the project pulls from the latest upstream premake5, however the build script shows it's not cloning to build:
> 
> ```
> error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
> error: 6059 bytes of body are still expected
> fetch-pack: unexpected disconnect while reading sideband packet
> fatal: early EOF
> fatal: fetch-pack: invalid index-pack output
> fatal: clone of 'https://github.com/premake/premake-core.git' into submodule path '/home/testing/Desktop/xenia/xenia-canary/third_party/premake-core' failed
> ```
> 
> Otherwise, `cmake 3.28.3`, `ninja 1.11.1` and clang/llvm `18.1.3` are installed.

Yeah, I've just tested my build script and premake cloned and built fine, so it might be just a network error while fetching data on your side.

Shoegzer commented:
Thanks. It's odd though, I don't get these kinds of errors pulling submodules with any other project. What are your platform specs?  I'm curious if others with a Ubuntu 24.04 base are encountering this issue.

Shoegzer commented:
Also: The xb script runs into problems cloning certain repositories:
```
Cloning into '/home/testing/Desktop/xenia-canary/third_party/premake-androidndk'...
fatal: unable to access 'https://github.com/Triang3l/premake-androidndk.git/': Could not resolve host: github.com
```
However, when I manually clone these repositories outside the xb script, everything is fine:
```
git clone https://github.com/Triang3l/premake-androidndk.git/
Cloning into 'premake-androidndk'...
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (11/11), done.
remote: Total 13 (delta 2), reused 13 (delta 2), pack-reused 0 (from 0)
Receiving objects: 100% (13/13), 55.04 KiB | 1.49 MiB/s, done.
Resolving deltas: 100% (2/2), done.
```

RodoMa92 commented:
> Thanks. It's odd though, I don't get these kinds of errors pulling submodules with any other project. What are your platform specs? I'm curious if others with a Ubuntu 24.04 base are encountering this issue.

Arch base here.
